src/hotspot/share/oops/klassVtable.cpp

```diff
@@ -222,13 +222,21 @@ void klassVtable::initialize_vtable(bool checkconstraints, TRAPS) {
           assert(def_vtable_indices->length() == len, "reinit vtable len?");
         }
         for (int i = 0; i < len; i++) {
-          methodHandle mh(THREAD, default_methods->at(i));
-          assert(!mh->is_private(), "private interface method in the default method list");
-          bool needs_new_entry = update_inherited_vtable(mh, super_vtable_len, i, checkconstraints, CHECK);
+          bool needs_new_entry;
+          {
+            // Reduce the scope of this handle so that it is fetched again.
+            // The methodHandle keeps it from being deleted by RedefineClasses while
+            // we're using it.
+            methodHandle mh(THREAD, default_methods->at(i));
+            assert(!mh->is_private(), "private interface method in the default method list");
+            needs_new_entry = update_inherited_vtable(mh, super_vtable_len, i, checkconstraints, CHECK);
+          }
 
           // needs new entry
           if (needs_new_entry) {
-            put_method_at(mh(), initialized);
+            // Refetch this default method in case of redefinition in safepoint above.
```

Comment:
Which code above can go to a safepoint? So perhaps reword the comment:

```
// Refetch this default method in case of redefinition in a safepoint that
// might happen in XXX above.
```

Reply:
// Refetch this default method in case of redefinition that might
            // happen during constraint checking in the update_inherited_vtable call above.
How about this?

Reply:
Looks good to me.